File: ruby/src/queries/accuracy_check_query.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'net/http'
require 'csv'
require 'json'
require_relative '../botpress_endpoint'

module Queries
  # Sends each question from a test-data CSV to a Botpress /converse endpoint
  # and returns the parsed response bodies.
  class AccuracyCheckQuery
    INVALID_PATTERNS = %r![\\'|`\^"<>)(}{\]\[;/?:@&=+$,%\# ]!

    # @rbs endpoint: BotpressEndpoint
    # @rbs path_to_test_data: String
    # @rbs return: Array[Hash[String, untyped]]
    def self.request!(endpoint:, path_to_test_data:)
      new(endpoint:, path_to_test_data:).request!
    end

    # @rbs endpoint: BotpressEndpoint
    # @rbs path_to_test_data: String
    # @rbs return: void
    def initialize(endpoint:, path_to_test_data:)
      @scheme    = endpoint.scheme.gsub(INVALID_PATTERNS, '')
      @host      = endpoint.host.gsub(INVALID_PATTERNS, '')
      @bot_id    = endpoint.bot_id.gsub(INVALID_PATTERNS, '')
      @user_id   = endpoint.user_id.gsub(INVALID_PATTERNS, '')
      @test_data = CSV.read(path_to_test_data, headers: true)
    end

    # @rbs return: Array[Hash[String, untyped]]
    def request!
      authorized
      request.map(&:body).map do |res_body|
        JSON.parse(res_body)
      end
    end

    private

    attr_reader :scheme, :host, :bot_id, :user_id, :test_data, :req

    # @rbs return: URI::Generic
    def uri
      @uri ||= URI.parse("#{scheme}://#{host}/api/v1/bots/#{bot_id}/converse/#{user_id}/secured?include=suggestions")
    end

    # @rbs return: void
    def authorized
      @req                 = Net::HTTP::Post.new(uri)
      @req[:authorization] = ENV.fetch('BOTPRESS_ACCESS_TOKEN', '')
    end

    # @rbs return: Array[Net::HTTPResponse]
    def request
      test_data['Question'].map do |question|
        req.set_form_data(type: :text, text: question)
        http_client.start { |http| http.request(req) }
      end
    end

    # @rbs return: Net::HTTP
    def http_client
      @http_client ||= Net::HTTP.new(uri.host.to_s, uri.port).tap do |client|
        client.use_ssl = true if uri.to_s.include?('https')
      end
    end
  end
end

File: python/src/queries/accuracy_check_query.py
import csv
import json
import os
import re
import urllib.request
from typing import Any
from urllib.parse import urlencode

from botpress_endpoint import BotpressEndpoint

INVALID_PATTERNS = r'[\\\'|`^"<>)(}{\]\[;/?:@&=+$,%# ]'


class AccuracyCheckQuery:
    """Sends each question from a test-data CSV to a Botpress /converse
    endpoint and returns the parsed response bodies."""

    @classmethod
    def request(
        cls,
        endpoint: BotpressEndpoint,
        path_to_test_data: str,
    ) -> list[dict[str, Any]]:
        return cls(endpoint, path_to_test_data)._request()

    def __init__(
        self,
        endpoint: BotpressEndpoint,
        path_to_test_data: str,
    ) -> None:
        self._scheme = re.sub(INVALID_PATTERNS, '', endpoint.scheme)
        self._host = re.sub(INVALID_PATTERNS, '', endpoint.host)
        self._bot_id = re.sub(INVALID_PATTERNS, '', endpoint.bot_id)
        self._user_id = re.sub(INVALID_PATTERNS, '', endpoint.user_id)
        with open(path_to_test_data) as f:
            self._test_data = list(csv.DictReader(f))

    # private

    def _request(self) -> list[dict[str, Any]]:
        return [json.loads(response.read()) for response in self._responses()]

    def _uri(self) -> str:
        return (
            f'{self._scheme}://{self._host}/api/v1/bots/{self._bot_id}'
            f'/converse/{self._user_id}/secured?include=suggestions'
        )

    def _responses(self) -> list[Any]:
        uri = self._uri()
        headers = {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Authorization': os.environ.get('BOTPRESS_ACCESS_TOKEN', ''),
        }
        return [
            urllib.request.urlopen(
                urllib.request.Request(
                    uri,
                    urlencode(
                        {'type': 'text', 'text': test_datum['Question']}
                    ).encode(),
                    headers,
                )
            )
            for test_datum in self._test_data
        ]
